File: sdk/transaction.py
"""XRP transaction builder for pointer submissions."""

from __future__ import annotations

from typing import Dict

from .pointer import encode_pointer_memo
from .validation import (
    validate_cid,
    validate_non_negative_int,
    validate_xrp_address,
    ValidationError,
)


DEFAULT_DESTINATION = "rwdm72S9YVKkZjeADKU2bbUMuY4vPnSfH7"
DEFAULT_MEMO_TYPE = "pf.ptr"
DEFAULT_AMOUNT_DROPS = "1"
MAX_MEMO_BYTES = 1024


def build_pointer_transaction(
    account: str,
    cid: str,
    kind: str = "TASK_SUBMISSION",
    destination: str = DEFAULT_DESTINATION,
    schema: int = 1,
    flags: int = 1,
    amount_drops: str = DEFAULT_AMOUNT_DROPS,
) -> Dict[str, object]:
    """Build an XRP Payment transaction with a pointer memo."""
    try:
        validate_xrp_address(account)
        validate_xrp_address(destination)
        validate_cid(cid)
        validate_non_negative_int(schema, "schema")
        validate_non_negative_int(flags, "flags")
    except ValidationError as exc:
        raise ValueError(str(exc)) from exc

    memo_payload = encode_pointer_memo(cid, kind=kind, schema=schema, flags=flags)
    if not amount_drops.isdigit() or int(amount_drops) <= 0:
        raise ValueError("amount_drops must be a positive integer string")
    if len(memo_payload) > MAX_MEMO_BYTES:
        raise ValueError(f"MemoData too large: {len(memo_payload)} bytes (max {MAX_MEMO_BYTES})")

    return {
        "TransactionType": "Payment",
        "Account": account,
        "Destination": destination,
        "Amount": amount_drops,
        "Memos": [
            {
                "Memo": {
                    "MemoType": DEFAULT_MEMO_TYPE.encode("utf-8").hex(),
                    "MemoData": memo_payload.hex(),
                }
            }
        ],
    }

File: ts/src/transaction.ts
import { encodePointerMemo } from "./pointer.js";
import {
  validateCid,
  validateNonNegativeInt,
  validateXrpAddress,
  ValidationError,
} from "./validation.js";

export const DEFAULT_DESTINATION = "rwdm72S9YVKkZjeADKU2bbUMuY4vPnSfH7";
export const DEFAULT_MEMO_TYPE = "pf.ptr";
export const DEFAULT_AMOUNT_DROPS = "1";
export const MAX_MEMO_BYTES = 1024;

export function buildPointerTransaction(
  account: string,
  cid: string,
  kind = "TASK_SUBMISSION",
  destination = DEFAULT_DESTINATION,
  schema = 1,
  flags = 1,
  amountDrops = DEFAULT_AMOUNT_DROPS
) {
  try {
    validateXrpAddress(account);
    validateXrpAddress(destination);
    validateCid(cid);
    validateNonNegativeInt(schema, "schema");
    validateNonNegativeInt(flags, "flags");
  } catch (err) {
    if (err instanceof ValidationError) throw err;
    throw err;
  }

  if (!/^\d+$/.test(amountDrops) || Number(amountDrops) <= 0) {
    throw new Error("amountDrops must be a positive integer string");
  }

  const memoPayload = encodePointerMemo(cid, kind, schema, flags);
  if (memoPayload.length > MAX_MEMO_BYTES) {
    throw new Error(`MemoData too large: ${memoPayload.length} bytes`);
  }

  return {
    TransactionType: "Payment",
    Account: account,
    Destination: destination,
    Amount: amountDrops,
    Memos: [
      {
        Memo: {
          MemoType: Buffer.from(DEFAULT_MEMO_TYPE, "utf8").toString("hex"),
          MemoData: Buffer.from(memoPayload).toString("hex"),
        },
      },
    ],
  };
}
